### DISCUSSION TITLE
Automated contract testing

### DISCUSSION THREAD

**Discussion by aaltw** [framna-nl-archive/krew-workspace#123] [Ideas]
Bij veel projecten hebben we vaak dat we tegen een API aan moeten praten. Gelukkig wordt die API door M4M zelf gemaakt, en gegenereerd met Open Api Spec (+ swagger achtige documentatie). Heel behulpzaam, maar we kunnen er ook voor zorgen dat we die zelf kunnen gebruiken.

Ik kwam de volgende package tegen (https://www.npmjs.com/package/openapi-typescript), daarmee kun je op basis van de API docs een setje interfaces maken. Dit kan een hoop typewerk schelen.

Wat ik zou willen hebben:
1. Script die automatisch die interfaces aanmaakt
2. Script die modificaties doorvoert aan interfaces waar wij die willen hebben
3. Script die een vergelijking maakt tussen wat wij van de API krijgen en wat we nu aan interfaces hebben, en dan een warning / error geeft als die niet gelijk lopen (en mogelijk dat we kunnen 'annotaten' dat het een future, depricated of een unexpected change is. 

Wat zijn jullie gedachten hierover? Is het nuttig dit te automatiseren?

**Comment by guidovanvilsteren** [comment]
Vooral punt 3 is iets waar ik momenteel bij Nutt tegenaan loop. De BE die velden verplaatst/verwijderd terwijl die nog wel doodleuk in de FE verwacht worden.
Kun je zoiets met een interval kunnen laten draaien? Eigenlijk wil je dat bij de BE ook telkens zo'n check gebeurt bij het aanmaken van een PR o.i.d.

**Comment by aaltw** [comment]
@cassshh jij bent hier toch ook iets voor aan het doen? Althans.. je was van openAPI spec naar .ts conversie aan het maken, toch?

**Reply by cassshh** [reply]
@aaltw enigszins.
Ik was aan het experimenteren met [OpenAPI Generator](https://github.com/OpenAPITools/openapi-generator).
Echter is dat puur genereren o.b.v. een OAS (OpenAPI Spec).

Maar als je meer kijkt richting je 3e punt;
Dat zou wellicht kunnen met iets als [Ajv JSON schema validator](https://github.com/ajv-validator/ajv)
